traffic_router/ultimate-test-harness/dns_test.go

```diff
@@ -0,0 +1,136 @@
+package ultimate_test_harness
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"text/tabwriter"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type DNSBenchmark struct {
+	Benchmark
+}
+
+func RunDNSBenchmarksAgainstTrafficRouters(t *testing.T, benchmark DNSBenchmark) {
+	passedTests := 0
+	failedTests := 0
+
+	fmt.Printf("Passing criteria: Routing at least %d requests per second\n", benchmark.RequestsPerSecondThreshold)
+	writer := tabwriter.NewWriter(os.Stdout, 20, 8, 1, '\t', tabwriter.AlignRight)
+	fmt.Fprintf(writer, "%s\t%s\t%s\t%s\t%s\t%s\t%s\n", "Traffic Router", "Protocol", "Delivery Service", "Passed?", "Requests Per Second", "Answers", "Failures")
+	for trafficRouterIndex, _ := range benchmark.TrafficRouters {
+		trafficRouter := &benchmark.TrafficRouters[trafficRouterIndex]
```

Comment:
nit: you can get this value from the `range` statement, that'll save you another lookup.

Reply:
The problem with getting `trafficRouter` from the `range` statement is that, since that `TRDetails`in `benchmark.TrafficRouters` is copied by value to `trafficRouter`, when `Run()` tries later to get that `TRDetails`, none of the changes made in this `for` loop are still there. So it needs to be gotten by reference.

Reply:
Made the `TrafficRouter` field an array of references in 450f186990